src/main.rs

```diff
@@ -37,7 +38,40 @@ struct MarketData {
 }
 
 impl MarketData {
-    fn print_holders_map(&self) {
+    fn validate_holders(&self, hash: &String) -> Vec<FileRequest> {
+        // make a map of holders already printed
+        let mut previous_holders: HashMap<&String, &FileRequest> = HashMap::new();
+
+        for holder in &self.files[hash] {
+            let user = &holder.user;
+            // check if the user has expired
+            if holder.expiration < get_current_time() {
+              continue;
+            } // this if statement must be first, otherwise it may unecessarily add expired holders or compare with expired holders
+            // if both duplicated holders are expired - we don't need either.
+            if previous_holders.contains_key(&user.id) {
+                // check which holder has the most recent ttl
+                // if the current holder has a more recent ttl, print it and remove the previous holder
+                // if the previous holder has a more recent ttl, skip the current holder
+                let prev_holder = previous_holders.get(&user.id).unwrap();
+                let current_holder_ttl = holder.expiration - get_current_time(); // current holder: smaller means older
+                let previous_holder_ttl: u64 = &prev_holder.expiration - get_current_time(); // previous holder: smaller means older
```

Comment:
The subtration of get_current_time() isn't necessary, you can just compare the two expirations as they are unix times